Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [9]:
import pandas as pd
import numpy as np

In [162]:
data = pd.read_csv('heart_pandas.csv')

In [161]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

In [207]:
## Ваш код
genders = data['sex'].value_counts()

male      207
female     96
Name: sex, dtype: int64


In [27]:
## Ваш ответ:
# male      207
# female     96

 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

In [138]:
## Ваш код
maleRatio = "{:.0%}".format(data['sex'].value_counts(normalize=True)[0])

In [23]:
## Ваш ответ: 68%

**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

In [208]:
## Ваш код
# print(data.groupby(['sex']))
heartDiseasesM = data[(data['sex'] == 'male') & (data['target'] == 1)].count()
heartDiseasesF = data[(data['sex'] == 'female') & (data['target'] == 1)].count()

In [94]:
## Ваш ответ: male: 93; female: 72;

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [117]:
## Ваш код
healthyMaleRatio = data[(data['sex'] == 'male') & (data['target'] == 0)].count()
print(healthyMaleRatio / data.count()[0])

In [89]:
## Ваш ответ: 0.376238

**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [214]:
## Ваш код
youngHDClient = data[data['target'] == 1]['age'].min()
print(youngHDClient)

29


In [215]:
## Ваш ответ: 29

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [216]:
## Ваш код
oldHealthyClient = data[data['target'] == 0]['age'].max()

In [217]:
## Ваш ответ: 77

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [212]:
## Ваш код
youngHDFemale = data[(data['target'] == 1) & (data['sex'] == 'female')]['age'].min()

In [22]:
## Ваш ответ: 34

**Какой средний возраст женщин?**

In [209]:
## Ваш код
femaleMean = data[data['sex'] == 'female']['age'].mean()

In [24]:
## Ваш ответ: 55.677

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [149]:
## Ваш код
hdMean = data[data['target'] == 1]['age'].mean()
hdStd = data[data['target'] == 1]['age'].std()

In [25]:
## Ваш ответ: mean = 52.5; std = 9.55;

** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

In [155]:
## Ваш код
healthyChol = data[data['target'] == 0]['cholesterol'].mean()
avChol = data['cholesterol'].mean()

if healthyChol < avChol:
    print(True)
else:
    print(False)

In [27]:
## Ваш ответ: Неправда

**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [174]:
## Ваш код
data.groupby(['rest_ecg'])[data.select_dtypes(include=[np.number]).columns.tolist()].agg([np.max, np.mean])

age            resting_blood_pressure  \
                             amax       mean                   amax   
rest_ecg                                                              
ST-T wave abnormality          71  52.914474                    180   
left ventricular hypertrophy   76  61.000000                    180   
normal                         77  55.687075                    200   

                                         cholesterol              \
                                    mean        amax        mean   
rest_ecg                                                           
ST-T wave abnormality         129.065789         354  237.269737   
left ventricular hypertrophy  140.500000         327  261.750000   
normal                        134.027211         564  255.142857   

                             max_heart_rate_achieved              \
                                                amax        mean   
rest_ecg                                                           
ST-T wave abnormality                            194  151.960526   
left ventricular hypertrophy                     140  125.750000   
normal                                           202  147.904762   

                             st_depression           num_major_vessels  \
                                      amax      mean              amax   
rest_ecg                                                                 
ST-T wave abnormality                  5.6  0.879605                 4   
left ventricular hypertrophy           4.4  2.725000                 3   
normal                                 6.2  1.159184                 4   

                                       target            
                                  mean   amax      mean  
rest_ecg                                                 
ST-T wave abnormality         0.638158      1  0.631579  
left ventricular hypertrophy  1.000000      1  0.250000  
normal                        0.816327      1  0.462585

In [35]:
## Ваш ответ:

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [167]:
## Ваш код
maleDepr = data[(data['sex'] == 'male') & (data['target'] == 1)]['st_depression'].mean()
femaleDepr = data[(data['sex'] == 'female') & (data['target'] == 0)]['st_depression'].mean()

if maleDepr > femaleDepr:
    print('Уровень депресси при физ нагрузке выше у больных мужчин')
else:
    print('Уровень депресси при физ нагрузке выше у здоровых женщин')


In [37]:
## Ваш ответ: Уровень депресси при физ нагрузке выше у здоровых женщин

**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [184]:
## Ваш код
data.groupby(['chest_pain_type', 'rest_ecg', 'thalassemia'])['cholesterol'].agg([np.max, np.min])

amax  amin
chest_pain_type  rest_ecg                     thalassemia                  
atypical angina  ST-T wave abnormality        fixed defect        340   141
                                              normal              218   218
                                              reversable defect   313   126
                 left ventricular hypertrophy fixed defect        197   197
                 normal                       fixed defect        417   149
                                              normal              256   216
                                              reversable defect   564   212
non-anginal pain ST-T wave abnormality        fixed defect        240   204
                                              reversable defect   298   199
                 normal                       fixed defect        283   182
                                              normal              233   186
                                              reversable defect   288   193
typical angina   ST-T wave abnormality        fixed defect        354   149
                                              normal              315   169
                                              reversable defect   353   131
                 left ventricular hypertrophy fixed defect        327   327
                                              normal              318   318
                                              reversable defect   205   205
                 normal                       fixed defect        394   166
                                              normal              276   176
                                              reversable defect   409   164

In [ ]:
## Ваш ответ:

**Сколько значений может принимать каждый из категориальных признаков?**

In [240]:
## Ваш код
data = data.replace({'target': {1: 'Sick', 0: 'Healthy'}})
categorical_columns = [c for c in data.columns if data[c].dtype.name == 'object']

for c in categorical_columns:
    print(c + ' ' + str(data[c].nunique()))

sex 2
chest_pain_type 3
fasting_blood_sugar 2
rest_ecg 3
exercise_induced_angina 2
st_slope 2
thalassemia 3
target 2


In [205]:
## Ваш ответ:
# sex 2
# chest_pain_type 3
# fasting_blood_sugar 2
# rest_ecg 3
# exercise_induced_angina 2
# st_slope 2
# thalassemia 3
# target 2

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [238]:
## Ваш код
categorical_columns = [c for c in data.columns if data[c].dtype.name == 'object']

d = {}

for c in categorical_columns:
    maxNum = data[c].value_counts().max()
    minNum = data[c].value_counts().min()
    
    d.update({c: (minNum / maxNum)})
    
print(sorted(d.items(), key=lambda x: x[1])[0])

('rest_ecg', 0.02631578947368421)


In [226]:
## Ваш ответ: 'rest_ecg'

---

@Booster
Мамкин Data Scientist

tg @mommyscience